libraries/botbuilder-core/botbuilder/core/user_state.py

```diff
@@ -0,0 +1,45 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+from .bot_context import BotContext
+from .bot_state import BotState
+from .storage import Storage
+
+
+class UserState(BotState):
+    """
+    Reads and writes user state for your bot to storage.
+    """
+
+    no_key_error_message = 'UserState: channel_id and/or conversation missing from context.activity.'
+
+    def __init__(self, storage: Storage, namespace=''):
+        """
+        Creates a new UserState instance.
+        :param storage:
+        :param namespace:
+        """
+        self.namespace = namespace
+
+        def call_get_storage_key(context):
+            key = self.get_storage_key(context)
+            if key is None:
+                raise AttributeError(self.no_key_error_message)
+            else:
+                return key
+
+        super(UserState, self).__init__(storage, call_get_storage_key)
+
+    def get_storage_key(self, context: BotContext) -> str:
+        """
+        Returns the storage key for the current user state.
+        :param context:
+        :return:
+        """
+        activity = context.activity
+        channel_id = activity.channel_id or None
+        user_id = activity.from_property.id or None
```

Comment:
check if from_property exists, would id throw exception?

Reply:
refactored this